src/prep.py

```diff
@@ -0,0 +1,365 @@
+
+    Returns the path to the .i file as well as the stdout from `make`
+    """
+
+    if (abs_target := path_to_linux / target).exists():
+        double_check_removal_with_user(
+            file=abs_target, force_rm=force_rm_existing_target
+        )
+
+    info(f"Making {target}...")
+    make_output = subprocess.run(
+        [*build_cmd, target], check=False, capture_output=True, cwd=path_to_linux
+    )
+    ic(make_output)
+
+    destination = output_dir / target.name
+    if make_output.returncode == 0:
+        info(f"Moving generated {target} file to {output_dir}")
+        shutil.move(path_to_linux / target, destination)
+        success(f"Successfully generated {target.name}")
+
+    return (destination, make_output.stdout.decode("utf-8"))
+
+
+def cleanup_dot_i_file(path: Path) -> None:
+    """
+    Remove preprocessor-specific lines to speed up reductions
+
+    Got some tips from here:
+    https://gcc.gnu.org/wiki/A_guide_to_testcase_reduction
+
+    The overhead of this function is worth it. Reductions are sped up greatly.
+    """
+    info(f"Cleaning up {path.name} file for reduction")
+    preprocess_pattern = r"^# .*$"
+
+    file_raw = path.read_text()
+    file_raw = re.sub(preprocess_pattern, "", file_raw, flags=re.MULTILINE)
+
+    path.write_text(file_raw)
+
+
+def get_compiler_invocation(target: Path, make_stdout: str) -> Command:
+    """
+    Parse out the CC invocation from make's raw output with V=1 on
+    """
+    pattern = f".*-o {target}.*"
+    matches = re.findall(pattern, make_stdout)
+    ic(matches)
+    assert (
+        len(matches) == 1
+    ), "Too many (or too few) compiler invocation matches!\n\
+            (not your fault, it's prepreduce's fault)"
+
+    main_invocation = matches[0].split()
+    main_invocation.insert(-1, "-c")
+    return clean_compiler_invocation(main_invocation)
+
+
+def clean_compiler_invocation(cc_invocation: Command) -> Command:
+    """
+    Remove all preprocessor flags as well as any relative pathing in invocation
+    """
+
+    to_remove = ("-I", "-D", "-Wp", "-include", "-Werror", "./", "-U", "-E")
+    rules = [(lambda rule: lambda flag: rule not in flag)(rule) for rule in to_remove]
+
+    cleaned = cc_invocation.copy()
+    for rule in rules:
+        cleaned = filter(rule, cleaned)
+
+    cleaned = list(cleaned)  # eagerly convert iterator to list
+
+    ic(cleaned)
+
+    return cleaned
+
+
+def cc_invocation_to_flags(cc_invocation: Command) -> str:
+    """
+    Accepts a full cc_invocation command and parses out just the flags
+
+    The order of the pop()'s are important as the relative index of items
+    changes as items are removed
+    """
+    cc_flags = cc_invocation[1:].copy()  # ditch the `clang` or `gcc` invocation
+
+    dash_o_idx = cc_flags.index("-o")
+    cc_flags.pop(dash_o_idx + 1)  # remove target.o
+    cc_flags.pop(dash_o_idx)  # remove -o
+
+    dash_c_idx = cc_flags.index("-c")
+    cc_flags.pop(dash_c_idx + 1)  # remove target.c
```

Comment:
careful, target.c doesn't necessarily follow `-c` like `target.o` follows `-o`. (`-c` is position independent relative to other args).

Do we have the target name say from perhaps the caller of `cc_invocation_to_flags`?

Reply:
Whoops, I naively assumed `-c` was positional.

Will fix.